NBXplorer/Repository.cs

```diff
@@ -1149,67 +1216,103 @@ public Task<TrackedTransaction[]> GetMatches(NBitcoin.Transaction tx, uint256 bl
+		{
+			var outPointWithScript = new List<(OutPoint, Script)>();
+			if (!txn.IsCoinBase)
+			{
+				foreach (var input in txn.Inputs)
+				{
+					outPointWithScript.Add((input.PrevOut, null));
+				}
+			}
+			foreach (var coin in txn.Outputs.AsCoins())
+			{
+				if (MinUtxoValue != null && coin.Amount < MinUtxoValue)
+					continue;
+				outPointWithScript.Add((coin.Outpoint, coin.ScriptPubKey));
+			}
+			return outPointWithScript;
+		}
+
+		private void recordMatches(
+									IReadOnlyCollection<NBitcoin.Transaction> txs,
+									IReadOnlyCollection<KeyPathInformation> keyPathInformations,
+									ref MultiValueDictionary<TrackedTransaction, KeyPathInformation> keyPathInformationsByTrackedTransaction,
+									ref HashSet<uint256> noMatchTransactions,
+									ref Dictionary<string, TrackedTransaction> matches,
+									uint256 blockId,
+									DateTimeOffset now
+									)
+		{
+			foreach (var tx in txs)
+			{
+				if (keyPathInformations.Count != 0)
+					noMatchTransactions.Remove(tx.GetHash());
+				foreach (var keyInfo in keyPathInformations)
+				{
+					var matchesGroupingKey = $"{keyInfo.DerivationStrategy?.ToString() ?? keyInfo.ScriptPubKey.ToHex()}-[{tx.GetHash()}]";
+					if (!matches.TryGetValue(matchesGroupingKey, out TrackedTransaction match))
+					{
+						match = CreateTrackedTransaction(keyInfo.TrackedSource,
+							new TrackedTransactionKey(tx.GetHash(), blockId, false),
+							tx,
+							new Dictionary<Script, KeyPath>());
+						match.FirstSeen = now;
+						match.Inserted = now;
+						matches.Add(matchesGroupingKey, match);
+					}
+					if (keyInfo.KeyPath != null)
+						match.KnownKeyPathMapping.TryAdd(keyInfo.ScriptPubKey, keyInfo.KeyPath);
+					keyPathInformationsByTrackedTransaction.Add(match, keyInfo);
+				}
+			}
+		}
 		public async Task<TrackedTransaction[]> GetMatches(IList<NBitcoin.Transaction> txs, uint256 blockId, DateTimeOffset now, bool useCache)
 		{
 			foreach (var tx in txs)
 				tx.PrecomputeHash(false, true);
 			var transactionsPerScript = new MultiValueDictionary<Script, NBitcoin.Transaction>();
+			var transactionsPerOutpoint = new MultiValueDictionary<OutPoint, NBitcoin.Transaction>();
 			var matches = new Dictionary<string, TrackedTransaction>();
-			HashSet<Script> scripts = new HashSet<Script>(txs.Count);
 			var noMatchTransactions = new HashSet<uint256>(txs.Count);
+			var outPointsWithScripts = new HashSet<(OutPoint outPoint, Script script)>();
 			foreach (var tx in txs)
 			{
 				if (blockId != null && useCache && noMatchCache.Contains(tx.GetHash()))
 				{
 					continue;
 				}
 				noMatchTransactions.Add(tx.GetHash());
-				foreach (var input in tx.Inputs)
-				{
-					var signer = input.GetSigner();
-					if (signer != null)
-					{
-						scripts.Add(signer.ScriptPubKey);
-						transactionsPerScript.Add(signer.ScriptPubKey, tx);
-					}
-				}
-				foreach (var output in tx.Outputs)
+				var txOutPointsWithScripts = ExtractInfo(tx);
```

Comment:
Since you use `ExtractInfo` only here, I would advise to just iterate through directly in this method, adding to `transactionsPerOutpoint` and `outPointsWithScripts` along the way.

Reply:
Fixed this